src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala

```diff
@@ -0,0 +1,387 @@
+      )
+        zeroKW
+      else
+        setPower
+
+    // net power after considering efficiency
+    val netPower = calcNetPower(adaptedSetPower)
+
+    val currentState =
+      StorageState(
+        currentStoredEnergy,
+        adaptedSetPower,
+        data.currentTick,
+      )
+
+    // if the storage is at minimum or maximum charged energy AND we are charging
+    // or discharging, flex options will be different at the next activation
+    val isEmptyOrFull =
+      isEmpty(currentStoredEnergy) || isFull(currentStoredEnergy)
+    // if target soc is enabled, we can also be at that exact point
+    val isAtTarget = refTargetSoc.exists { targetParams =>
+      currentStoredEnergy <= targetParams.targetWithPosMargin &&
+      currentStoredEnergy >= targetParams.targetWithNegMargin
+    }
+    val isChargingOrDischarging = netPower != zeroKW
+    // if we've been triggered just before we hit the minimum or maximum energy,
+    // and we're still discharging or charging respectively (happens in edge cases),
+    // we already set netPower to zero (see above) and also want to refresh flex options
+    // at the next activation.
+    // Similarly, if the ref target margin area is hit before hitting target SOC, we want
+    // to refresh flex options.
+    val hasObsoleteFlexOptions =
+      (isFull(currentStoredEnergy) && setPower > zeroKW) ||
+        (isEmpty(currentStoredEnergy) && setPower < zeroKW) ||
+        (isAtTarget && setPower != zeroKW)
+
+    val activateAtNextTick =
+      ((isEmptyOrFull || isAtTarget) && isChargingOrDischarging) || hasObsoleteFlexOptions
+
+    // calculate the time span until we're full or empty, if applicable
+    val maybeTimeSpan =
+      if (!isChargingOrDischarging) {
+        // we're at 0 kW, do nothing
+        None
+      } else if (netPower > zeroKW) {
+        // we're charging, calculate time until we're full or at target energy
+
+        val closestEnergyTarget = refTargetSoc
+          .flatMap { targetParams =>
+            Option.when(
+              currentStoredEnergy <= targetParams.targetWithNegMargin
+            )(targetParams.targetSoc)
+          }
+          .getOrElse(eStorage)
+
+        val energyToFull = closestEnergyTarget - currentStoredEnergy
+        Some(energyToFull / netPower)
+      } else {
+        // we're discharging, calculate time until we're at lowest energy allowed or at target energy
+
+        val closestEnergyTarget = refTargetSoc
+          .flatMap { targetParams =>
+            Option.when(
+              currentStoredEnergy >= targetParams.targetWithPosMargin
+            )(targetParams.targetSoc)
+          }
+          .getOrElse(minEnergy)
+
+        val energyToEmpty = currentStoredEnergy - closestEnergyTarget
+        Some(energyToEmpty / (netPower * (-1)))
+      }
+
+    // calculate the tick from time span
+    val maybeNextTick = maybeTimeSpan.map { timeSpan =>
+      val ticksToEmpty = Math.round(timeSpan.toSeconds)
```

Comment:
Wouldn't it be ticks to empty or full or target / limit?

Reply:
Good catch